app/Bundle.php

```diff
@@ -53,9 +57,100 @@ class Bundle extends Model
     protected $dates = [
         'created_at',
         'updated_at',
-        'allocated_at'  // When it was handed out.
+        'disbursed_at'  // When it was handed out.
     ];
 
+    /**
+     * Refactored function that works out if we broke anything then adds vouchers.
+     *
+     * @param Collection $vouchers
+     * @param array $codes
+     * @param Bundle|null $bundle
+     * @return array
+     */
+    private function alterVouchers(Collection $vouchers, array $codes, Bundle $bundle = null)
+    {
+        $errors = [];
+        $badCodes = array_diff($vouchers->pluck("code")->toArray(), $codes);
+
+        if (!isEmpty($badCodes)) {
+            //Stop! report the bad codes!
+            $errors["codes"] = (array_key_exists("badCodes", $errors))
+                ? array_merge($badCodes, $errors)
+                : $badCodes;
+        } else {
+            // Run all these.
+            $vouchers->each(
+                function (Voucher $voucher) use ($bundle, $errors) {
+                    try {
+                        $voucher->setBundle($bundle);
```

Comment:
`$bundle` might be null here is it ok to set that?

Reply:
Yes. If it;s null, we've put a null bundle, so set the voucher's bundle_id to null (disassociate it)
voucher->setBundle also tries to rollback the transition